app/api/entities/managerFunctions.ts

```diff
@@ -0,0 +1,219 @@
+import { groupBy } from 'lodash';
+import { WithId } from 'api/odm';
+import { attachmentsPath, uploadsPath, files as filesAPI, storeFile } from 'api/files';
+import { processDocument } from 'api/files/processDocument';
+import { search } from 'api/search';
+import { errorLog } from 'api/log';
+import { prettifyError } from 'api/utils/handleError';
+import { FileType } from 'shared/types/fileType';
+import { MetadataObjectSchema } from 'shared/types/commonTypes';
+import { EntityWithFilesSchema } from 'shared/types/entityType';
+import { ClientEntitySchema } from 'app/istore';
+
+const prepareNewFiles = async (
+  entity: EntityWithFilesSchema,
+  updatedEntity: EntityWithFilesSchema,
+  newAttachments: FileType[] = [],
+  newDocuments: FileType[] = []
+) => {
+  const attachments: FileType[] = [];
+  const documents: FileType[] = [];
+
+  const newUrls = entity.attachments?.filter(attachment => !attachment._id && attachment.url);
+
+  if (newAttachments.length) {
+    await Promise.all(
+      newAttachments.map(async (file: FileType) => {
+        const savedFile = await storeFile(attachmentsPath, file, true);
+        attachments.push({
+          ...savedFile,
+          entity: updatedEntity.sharedId,
+          type: 'attachment',
+        });
+      })
+    );
+  }
+
+  if (newDocuments.length) {
+    await Promise.all(
+      newDocuments.map(async (document: FileType) => {
+        const savedDocument = await storeFile(uploadsPath, document, true);
+        documents.push({
+          ...savedDocument,
+          entity: updatedEntity.sharedId,
+          type: 'document',
+        });
+      })
+    );
+  }
+
+  if (newUrls && newUrls.length) {
+    newUrls.map(async (url: any) => {
+      attachments.push({
+        ...url,
+        entity: updatedEntity.sharedId,
+        type: 'attachment',
+      });
+    });
+  }
+
+  return { attachments, documents };
+};
+
+const updateDeletedFiles = async (
+  entityFiles: WithId<FileType>[],
+  entity: EntityWithFilesSchema,
+  type: 'attachment' | 'document'
+) => {
+  const deletedFiles = entityFiles.filter(
+    existingFile =>
+      existingFile._id &&
+      existingFile.type === type &&
+      !entity[type === 'attachment' ? 'attachments' : 'documents']?.find(
+        attachment => attachment._id?.toString() === existingFile._id.toString()
+      )
+  );
+  await Promise.all(
```

Comment:
filesApi.delete works with a query, consider building one query (out of the _id's, perhaps?) instead of a Promise.all, to have less db calls.

Reply:
Nice catch! I'll check it